File: packages/js/core/src/algorithms/apply-redirects.ts
import { Uri, IUriRedirect } from "../types";

import { Tracer } from "@polywrap/tracing-js";
import { Result, ResultErr, ResultOk } from "@polywrap/result";

export const applyRedirects = Tracer.traceFunc(
  "core: applyRedirects",
  (uri: Uri, redirects: readonly IUriRedirect<Uri>[]): Result<Uri, Error> => {
    // Keep track of past redirects (from -> to) to find the final uri
    const redirectFromToMap: Record<string, Uri> = {};

    const createError = (message: string) => {
      const error = Error(
        `${message}\nResolution Stack: ${JSON.stringify(
          redirectFromToMap,
          null,
          2
        )}`
      );
      return ResultErr(error);
    };

    for (const redirect of redirects) {
      if (!redirect.from) {
        return createError(
          `Redirect missing the from property.\nEncountered while resolving ${uri.uri}`
        );
      }

      if (redirectFromToMap[redirect.from.uri]) {
        continue;
      }

      redirectFromToMap[redirect.from.uri] = redirect.to;
    }

    let finalUri = uri;

    const visitedUris: Record<string, boolean> = {};

    while (redirectFromToMap[finalUri.uri]) {
      visitedUris[finalUri.uri] = true;

      finalUri = redirectFromToMap[finalUri.uri];

      if (visitedUris[finalUri.uri]) {
        return createError(`Infinite loop while resolving URI "${uri}".`);
      }
    }

    return ResultOk(finalUri);
  }
);

File: packages/py/core/src/algorithms/apply_redirects.py
from json import dumps
from typing import Dict, List

from ..types import Uri, UriRedirect


def apply_redirects(uri: Uri, redirects: List[UriRedirect]) -> Uri:
    """
    Keep track of past redirects (from_uri -> to_uri) to find the final uri.

    :param Uri uri:
    :param List[UriRedirect] redirects:
    :return Uri:
    """

    def _throw_error(message: str, redirect_from_to_map: Dict[str, Uri]):
        serialize_message = {}
        for key in redirect_from_to_map:
            serialize_message[key] = redirect_from_to_map[key].uri
        raise ValueError(f"""{message}\nResolution Stack: {dumps(serialize_message)}""")

    redirect_from_to_map = {}
    for redirect in redirects:
        if not redirect.from_uri:
            _throw_error(
                f"Redirect missing the from_uri property.\nEncountered while resolving {uri.uri}",
                redirect_from_to_map,
            )

        if redirect_from_to_map.get(redirect.from_uri.uri):
            continue

        redirect_from_to_map[redirect.from_uri.uri] = redirect.to_uri

    final_uri = uri

    visited_uris = {}
    while redirect_from_to_map.get(final_uri.uri):
        visited_uris[final_uri.uri] = True
        final_uri = redirect_from_to_map[final_uri.uri]

        if visited_uris.get(final_uri.uri):
            _throw_error(f"Infinite loop while resolving URI '{uri}'.", redirect_from_to_map)
    return final_uri
